### PR TITLE
Add talisman api storage

### PR DESCRIPTION
**Pull request by Hat331**
add Talisman API REST-storage

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
 		"renderjson": "^1.4.0",
 		"spark-md5": "^3.0.1",
 		"sugar": "^1.5.0",
+		"uuid": "^8.3.2",
```
Comment (Yatskov) [inline package.json:36]: Still need this?

File: src/background/background.js [hunk 2]
```diff
@@ -15,19 +17,53 @@ config.loadConfiguration().then(() => {
...
+				});
+				let tToken = loginStatus.data.access_token;
+				store.axiosInstance.defaults.headers['Authorization'] = 'Bearer ' + tToken;
+			}
 		} else {
 			store = new StorePouchDB(config);
 		}
+
+		function _handleResponse(message) {
+			alert(`Message from the background script:  ${message.response}`);
```
Comment (Yatskov) [inline src/background/background.js:61]: i18n message?

File: src/background/background.js [hunk 3]
```diff
@@ -49,7 +85,21 @@ const sendToActiveTab = function (request, callback) {
...
-
+	if (request.logOut) {
+		if (
+			store.constructor.name === 'StoreTalismanApi' &&
+			request.logOut.url === store.axiosInstance.defaults.baseURL
+		) {
+			delete store.axiosInstance.defaults.headers.Authorization;
+			await store.axiosInstance.get('/oauth/logout');
+		} else {
+			console.log('IN ELSE LO');
```
Comment (Yatskov) [inline src/background/background.js:96]: debug not deleted?

File: src/_locales/ru/messages.json [hunk 4]
```diff
@@ -183,9 +183,31 @@
...
 	"options_rest_api_storage": { "message": "Серверное хранилище" },
+
+	"options_talisman_api_storage": { "message": "Хранилище Талисман" },
+	"options_talisman_url": { "message": "Адрес сервера Талисман" },
+	"options_talisman_login": { "message": "Адрес сервера Талисман" },
+	"options_talisman_password": { "message": "Пароль" },
+	"options_talisman_password_show_checkbox": { "message": " - Показать пароль" },
+	"options_talisman_url_placeholder": { "message": "Адрес сервера Талисман1" },
+	"options_talisman_login_placeholder": { "message": "Логин1" },
+	"options_talisman_password_placeholder": { "message": "Пароль1" },
```
Comment (Yatskov) [inline src/_locales/ru/messages.json:194]: why 1?

File: src/scripts/Config.js [hunk 5]
```diff
@@ -47,10 +65,11 @@ export default class Config {
 
 	/**
 	 * Saves configuration to chrome extension sync storage
-	 * @param {type} items
+	 * @param {{sitemapDb: string, restUrl: string, credential: string, storageType: (*|jQuery), dataDb: string, talismanApiUrl: string}} items
 	 * @returns {Promise<void>} Promise
 	 */
 	updateConfiguration(items) {
+		//	browser.storage.sync.get().then(result => {console.log(result);});111111111111111
```
Comment (Yatskov) [inline src/scripts/Config.js:72]: forgot to delete?

File: src/_locales/en/messages.json [hunk 6]
```diff
@@ -176,9 +178,24 @@
...
+	"options_talisman_api_storage": { "message": "Talisman Api" },
+	"options_talisman_url": { "message": "Talisman server URL" },
+	"options_talisman_login": { "message": "Login" },
+	"options_talisman_password": { "message": "Password" },
+	"options_talisman_password_show_checkbox": { "message": " - Show password" },
+	"options_talisman_url_placeholder": { "message": "Talisman server URL" },
+	"options_talisman_login_placeholder": { "message": "Login" },
+	"options_talisman_password_placeholder": { "message": "Password" },
+	"options_talisman_url_popup_title": { "message": "Talisman api url." },
+	"options_talisman_url_popup_content": { "message": "Sitemaps will be syncing with that URL." },
```
Comment (Yatskov) [inline src/_locales/en/messages.json:191]: Will "use this talisman stand"?

File: src/scripts/Controller.js [hunk 7]
```diff
@@ -643,20 +671,58 @@ export default class SitemapController {
...
-			const $sitemap = ich.SitemapListItem(sitemap);
-			$sitemap.data('sitemap', sitemap);
-			$sitemapListPanel.find('tbody').append($sitemap);
-		});
-		$('#viewport').html($sitemapListPanel);
-		Translator.translatePage();
+		if (sitemaps.error_msg) {
+			$('#sitemaps').hide();
+			$('#viewport').html(
+				'<div>Error! Please, check your login/pass, or connection availability!</div>'
```
Comment (Yatskov) [inline src/scripts/Controller.js:713]: Seems this was somewhere in translations

File: src/scripts/Controller.js [hunk 8]
```diff
@@ -1653,7 +1748,7 @@ export default class SitemapController {
 				if (url && attachments.has(url)) {
 					const attachment = { [selector.getUrlColumn()]: url };
 					Object.entries(attachments.get(url)).forEach(([key, value]) => {
-						attachment[`${selector.id}-${key}`] = value;
+						attachment[`${selector.uuid}-${key}`] = value;
```
Comment (Yatskov) [inline src/scripts/Controller.js:1751]: Why use uuid here?

File: src/scripts/Controller.js [hunk 9]
```diff
@@ -16,6 +16,16 @@ import SelectorTable from './Selector/SelectorTable';
 import Model from './Model';
 import Translator from './Translator';
 
+import { v4 as uuidv4 } from 'uuid';
```
Comment (Yatskov) [inline src/scripts/Controller.js:19]: not used?

File: src/scripts/Controller.js [hunk 10]
```diff
@@ -16,6 +16,16 @@ import SelectorTable from './Selector/SelectorTable';
...
+
+import {
+	authorizationFormInit,
+	checkTLogin,
+	tAuthFormSubmit,
+	tLogOut,
+} from './TalismanAuthorization';
+import { right } from 'core-js/internals/array-reduce';
```
Comment (Yatskov) [inline src/scripts/Controller.js:27]: not used?

File: src/scripts/Controller.js [hunk 11]
```diff
@@ -643,20 +671,58 @@ export default class SitemapController {
 		window.getSelection().removeAllRanges();
 	}
 
+	async authPage() {
+		$('.nav.navbar-nav').addClass('invisible');
+		$('#talisman-user-name').hide();
+		$('#talisman-logout-nav-button').hide();
```
Comment (Yatskov) [inline src/scripts/Controller.js:677]: You can reuse this jquery elements

File: src/scripts/Controller.js [hunk 12]
```diff
@@ -643,20 +671,58 @@ export default class SitemapController {
...
+
+		let request = {
+			storageType: true,
+		};
+		let storageType = await browser.runtime.sendMessage(request);
```
Comment (Yatskov) [inline src/scripts/Controller.js:682]: Why not move this check to StoreDevtools. And check it from store method?

File: src/scripts/Controller.js [hunk 13]
```diff
@@ -715,11 +781,40 @@ export default class SitemapController {
 		} else {
 			let sitemap = Sitemap.sitemapFromObj(sitemapObj);
 			sitemap._id = id;
+
+			if (!sitemap.new_version) {
+				this.transformOldToNew(sitemap);
+			}
+			sitemap.new_version = true;
```
Comment (Yatskov) [inline src/scripts/Controller.js:788]: Merge problems?

File: src/scripts/Selector.js [hunk 14]
```diff
@@ -50,8 +50,8 @@ export default class Selector {
 		if (Array.isArray(data)) {
 			return data.map(e => this.manipulateData(e));
 		}
-		if (Object.isObject(data) && this.id in data) {
-			data[this.id] = this.manipulateData(data[this.id]);
+		if (Object.isObject(data) && this.uuid in data) {
```
Comment (Yatskov) [inline src/scripts/Selector.js:53]: Merge problems?

File: src/scripts/SelectorList.js [hunk 15]
```diff
@@ -258,7 +258,7 @@ export default class SelectorList extends Array {
 	getSelectorById(selectorId) {
 		for (let i = 0; i < this.length; i++) {
 			const selector = this[i];
-			if (selector.id === selectorId) {
+			if (selector.uuid === selectorId) {
```
Comment (Yatskov) [inline src/scripts/SelectorList.js:261]: Merge problems?

File: yarn.lock [hunk 16]
```diff
@@ -1,9442 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (Yatskov) [inline yarn.lock:1]: Why deleted lock file

File: src/background/background.js [hunk 17]
```diff
@@ -48,26 +61,73 @@ const sendToActiveTab = function (request, callback) {
...
+				store.postInit(loginStatus.data.access_token);
+				let tToken = loginStatus.data.access_token;
+				store.axiosInstance.defaults.headers['Authorization'] = 'Bearer ' + tToken;
+				return {
+					authStatus: {
+						success: true,
+						username: request.credential.username,
+					},
+				};
+			}
```
Comment (mxsnq) [inline src/background/background.js:88]: move to store implementation

Comment (Hat331) [reply src/background/background.js]: Done

File: src/scripts/Controller.js [hunk 18]
```diff
@@ -643,20 +652,96 @@ export default class SitemapController {
 		window.getSelection().removeAllRanges();
 	}
 
+	async showAuthPage() {
+		let tUserName = $('#user-name');
+		let tLogoutButton = $('#logout-nav-button');
+
+		$('.nav.navbar-nav').addClass('invisible');
+		tUserName.hide();
+		tLogoutButton.hide();
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: Done

File: src/scripts/Sitemap.js [hunk 19]
```diff
@@ -19,7 +19,7 @@ export default class Sitemap {
 			sitemapObj._id,
 			sitemapObj.startUrls,
 			sitemapObj.model,
-			sitemapObj.selectors
+			sitemapObj.selectors || [] //if loaded sitemap not contained field "selectors", it can't be parsed by web-scrapper
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Sitemap.js:22]: Done

File: src/scripts/StoreTalismanApi.js [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+			bodyForm,
+			{
+				headers: {
+					'Content-Type': 'multipart/form-data',
+				},
+			}
+		);
+	}
+
+	postInit(tToken) {
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/StoreTalismanApi.js]: Done

File: yarn.lock (hunk 16 above)
Comment (Hat331) [reply yarn.lock]: Return it

File: src/scripts/Sitemap.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (mxsnq) [reply src/scripts/Sitemap.js:22]: Still there

**Comment by mxsnq** [comment]
merge master into current branch

File: src/scripts/StoreRestApi.js [hunk 21]
```diff
@@ -1,15 +1,20 @@
...
 		this.axiosInstance = axios.create({
-			baseURL: config.restUrl,
+			baseURL: baseUrl,
 		});
 		this.axiosInstance.defaults.headers.post['Content-Type'] = 'application/json';
 		this.axiosInstance.defaults.headers.put['Content-Type'] = 'application/json';
+		this.sitemapsPath = sitemapsPath;
+	}
+
+	postInit() {
```
Comment (mxsnq) [inline src/scripts/StoreRestApi.js:17]: I don't like that this method is called at different time for talisman and rest. I suggest either
- rename it to what it actually does (sets up response interceptors)
- unify response interception so it works for both storages: for example, check content type only for sitemap responses 
- move content type check to response handlers in other methods explicitly

File: src/scripts/Config.js [hunk 22]
```diff
@@ -47,7 +65,7 @@ export default class Config {
 
 	/**
 	 * Saves configuration to chrome extension sync storage
-	 * @param {type} items
+	 * @param {{sitemapDb: string, restUrl: string, credential: string, storageType: (*|jQuery), dataDb: string, talismanApiUrl: string}} items
```
Comment (mxsnq) [inline src/scripts/Config.js:68]: jQuery?

File: src/options/options.js [hunk 23]
```diff
@@ -55,12 +66,19 @@ function initConfigSwitch() {
...
 			$('.form-group.couchdb').hide();
+			$('.form-group.talisman').hide();
+		} else if (type === 'talisman') {
+			$('.form-group.rest').hide();
+			$('.form-group.couchdb').hide();
+			$('.form-group.talisman').show();
 		} else {
 			$('.form-group.rest').hide();
 			$('.form-group.couchdb').hide();
+			$('.form-group.talisman').hide();
```
Comment (mxsnq) [inline src/options/options.js:81]: hide all first, then show for given type

File: src/background/background.js [hunk 24]
```diff
@@ -48,26 +63,49 @@ const sendToActiveTab = function (request, callback) {
...
+	}
+
+	if (request.login) {
+		return await store.initTalismanLogin(request.credential).catch(er => {
+			return er;
+		});
+	}
+
+	if (request.logOut) {
+		await store.logOut();
```
Comment (mxsnq) [inline src/background/background.js:77]: return is missing

File: src/background/background.js [hunk 25]
```diff
@@ -48,26 +63,49 @@ const sendToActiveTab = function (request, callback) {
...
+	}
+
+	if (request.isAuthorized) {
+		const storeData = await store.isAuthorized();
+		return storeData ? { data: storeData } : false;
+	}
```
Comment (mxsnq) [inline src/background/background.js:83]: Splitting StoreDevTools in two feels weird to me with all their requests mixed here. Maybe we should merge them back? Additionally I'd rename talisman related requests to include talisman in their names (e.g. talismanLogin, taismanLogout, getTalismanUser)

File: src/devtools/views/AuthorizationPage.html [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+	#error {
+		background-color: #f44336;
+	}
+
+	#success {
+		background-color: #4caf50;
+	}
+</style>
+<!--ДОБАВИТЬ ПЕРЕВОД-->
+<center><h1>Authorization Page</h1></center>
```
Comment (mxsnq) [inline src/devtools/views/AuthorizationPage.html:18]: please do

File: src/devtools/views/AuthorizationPage.html [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+			></label>
+			<div class="col-lg-10">
+				<input
+					type="text"
+					class="form-control"
+					name="userLogin"
+					id="userLogin"
+					placeholder="authorization_login_placeholder"
+				/>
+			</div>
```
Comment (mxsnq) [inline src/devtools/views/AuthorizationPage.html:36]: Can we make these inputs shorter? Similar to selector edit panel.

**Review by mxsnq — APPROVED** [review]

File: src/background/background.js [hunk 28]
```diff
@@ -48,26 +63,49 @@ const sendToActiveTab = function (request, callback) {
...
 
 browser.runtime.onMessage.addListener(async request => {
-	console.log('browser.runtime.onMessage', request);
+	if (request.getStorageType) {
+		return store.constructor.name;
+	}
+
+	if (request.login) {
+		return await store.initTalismanLogin(request.credential).catch(er => {
+			return er;
```
Comment (mxsnq) [inline src/background/background.js:72]: this is dangerous

File: src/scripts/StoreTalismanApi.js [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+		super(config, baseUrl, sitemapsPath);
+	}
+
+	async initTalismanLogin(credentials) {
+		let bodyForm = new FormData();
+		let tLogin = credentials.username;
+		let tPassword = credentials.password;
+		bodyForm.append('username', tLogin);
+		bodyForm.append('password', tPassword);
+		let loginStatus = await this.axiosInstance.post(
```
Comment (mxsnq) [inline src/scripts/StoreTalismanApi.js:17]: Throws error on invalid credentials which replaces response in background.js and is not handled correctly.

**Review by Yatskov — APPROVED** [review]

**Pull request merged**